### PR TITLE
drivers: sensor: Add driver for TI INA3221

### PR DESCRIPTION
**Pull request by maxd-nordic**
This patch adds support for the TI INA3221 current monitor. This is the datasheet used for reference:
https://www.ti.com/lit/gpn/ina3221

Since this device has three channels, there is a custom attribute to select which channel is to be used when getting a sample. Measurements are done on all enabled channels.

### CODE REVIEW

File: drivers/sensor/ina3221/ina3221.c [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+			return ret;
+		}
+		k_sleep(K_USEC(avg_mode_samples[cfg->avg_mode] *
+			       conv_time_us[MAX(cfg->conv_time_shunt, cfg->conv_time_bus)]));
+	} else {
+		return -ENOTSUP;
+	}
+
+	while (!measurement_ready(dev)) {
+		k_sleep(K_USEC(100));
```
Comment (MaureenHelm) [inline drivers/sensor/ina3221/ina3221.c:178]: Why did you choose this sleep duration?

File: drivers/sensor/ina3221/ina3221.c [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+		.bus = I2C_DT_SPEC_INST_GET(index),                                                \
+		.avg_mode = DT_INST_PROP(index, avg_mode),                                         \
+		.conv_time_bus = DT_INST_PROP(index, conv_time_bus),                               \
+		.conv_time_shunt = DT_INST_PROP(index, conv_time_shunt),                           \
+		.enable_channel = DT_INST_PROP(index, enable_channel),                             \
+		.shunt_r = DT_INST_PROP(index, shunt_resistors),                                   \
+	};                                                                                         \
+	static struct ina3221_data ina3221_data_##index;                                           \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(index, ina3221_init, NULL, &ina3221_data_##index,                    \
```
Comment (MaureenHelm) [inline drivers/sensor/ina3221/ina3221.c:270]: `SENSOR_DEVICE_DT_INST_DEFINE`

File: dts/bindings/sensor/ti,ina3221.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+description: Texas Instruments INA3221 Triple-Channel Current/Power Monitor
+
+compatible: "ti,ina3221"
+
+include: [sensor-device.yaml, i2c-device.yaml]
+
+properties:
+  conv-time-shunt:
+    type: int
+    default: 4
```
Comment (MaureenHelm) [inline dts/bindings/sensor/ti,ina3221.yaml:13]: The default value needs a justification in the description. Is it a register reset value?

File: dts/bindings/sensor/ti,ina3221.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      - 1
+      - 2
+      - 3
+      - 4
+      - 5
+      - 6
+      - 7
+  conv-time-bus:
+    type: int
+    default: 4
```
Comment (MaureenHelm) [inline dts/bindings/sensor/ti,ina3221.yaml:35]: The default value needs a justification in the description. Is it a register reset value?

File: dts/bindings/sensor/ti,ina3221.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      - 1
+      - 2
+      - 3
+      - 4
+      - 5
+      - 6
+      - 7
+  avg-mode:
+    type: int
+    default: 0
```
Comment (MaureenHelm) [inline dts/bindings/sensor/ti,ina3221.yaml:57]: The default value needs a justification in the description. Is it a register reset value?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add this sensor to `tests/drivers/build_all/sensor`

File: drivers/sensor/ina3221/ina3221.c [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+	/* select first enabled channel */
+	for (size_t i = 0; i < 3; ++i) {
+		if (cfg->enable_channel[i]) {
+			data->selected_channel = i;
+			break;
+		}
+	}
+
+	/* check bus */
+	if (!device_is_ready(cfg->bus.bus)) {
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply drivers/sensor/ina3221/ina3221.c:65]: done

File: drivers/sensor/ina3221/ina3221.c (hunk 2 above)
Comment (maxd-nordic) [reply drivers/sensor/ina3221/ina3221.c:270]: done

File: dts/bindings/sensor/ti,ina3221.yaml (hunk 3 above)
Comment (maxd-nordic) [reply dts/bindings/sensor/ti,ina3221.yaml:13]: yes, added that in the description

File: dts/bindings/sensor/ti,ina3221.yaml (hunk 4 above)
Comment (maxd-nordic) [reply dts/bindings/sensor/ti,ina3221.yaml:35]: same as above

File: dts/bindings/sensor/ti,ina3221.yaml (hunk 5 above)
Comment (maxd-nordic) [reply dts/bindings/sensor/ti,ina3221.yaml:57]: same as above

File: dts/bindings/sensor/ti,ina3221.yaml [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      - 6
+      - 7
+  enable-channel:
+    type: array
+    default: [1, 1, 1]
+    description: |
+      List of flags to enable or disable each of the three channels.
+  shunt-resistors:
+    type: array
+    default: [1000, 1000, 1000]
```
[the remark this answers is not in the document]
Comment (maxd-nordic) [reply dts/bindings/sensor/ti,ina3221.yaml:84]: removed that default now.

**Review by MaureenHelm — DISMISSED** [review]
Needs rebase

**Comment by maxd-nordic** [comment]
> Needs rebase

done

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 8]
```diff
@@ -699,4 +699,13 @@ test_i2c_s11059: s11059@6b {
 	compatible = "hamamatsu,s11059";
 	reg = <0x6b>;
 	integration-time = <546000>;
+
+ina3221: ina3221@6c {
+	compatible = "ti,ina3221";
+	reg = <0x6c>;
+	shunt-resistors = < 1000 1000 1000 >;
```
Comment (cfriedt) [inline tests/drivers/build_all/sensor/i2c.dtsi:706]: @maxd-nordic - looks like spacing might be causing dtc compile issues. Can you try to remove the space after `<` and before `>`?

Comment (MaureenHelm) [reply tests/drivers/build_all/sensor/i2c.dtsi:706]: There's a missing curly brace closing the previous node

**Comment by maxd-nordic** [comment]
@MaureenHelm @cfriedt Could you have another look? It seems like I have to rebase daily because of new I2C devices.

**Review by MaureenHelm — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**